File: 3.longest-substring-without-repeating-characters/main.go
package main

import (
	"fmt"
)

func lengthOfLongestSubstring(s string) int {
	rk := 0
	nums := 0
	caps := map[rune]int{}
	for k := range s {
		if k != 0 {
			delete(caps, rune(s[k-1]))
		}
		for {
			if rk >= len(s) {
				break
			}
			if _, ok := caps[rune(s[rk])]; !ok {
				caps[rune(s[rk])] = 1
				rk++
			} else {
				break
			}
		}
		nums = Max(nums, rk-k)
	}
	return nums
}
func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {
	c := lengthOfLongestSubstring("pa")
	fmt.Println(c)
}

File: 3.longest-substring-without-repeating-characters/main.py
# coding: utf-8

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        m = 0
        for i in range(len(s)):
            hash = set()
            if s[i] not in hash:
                k = i
                while True:
                    if k < len(s) and (s[k] not in hash) :
                        hash.add(s[k])
                        k = k + 1
                    else:
                        m = max(m, k-i)
                        hash.remove(s[i])
                        break
        return m


if __name__ == "__main__":
    s = Solution()
    print("r:",s.lengthOfLongestSubstring('pwwkew'))
